src/statsmodels/contrasts.jl

```diff
@@ -0,0 +1,143 @@
+## Specify contrasts for coding categorical data in model matrix. Contrast types
+## are constructed with data and other options, creating a subtype of
+## AbstractContrast. AbstractContrast provides the interface for creating modle
+## matrix columns and coefficient names
+##
+## ModelFrame will hold a Dict{Symbol, T<:AbstractContrast} that maps column
+## names to contrasts.
+##
+## ModelMatrix will check this dict when evaluating terms, falling back to a
+## default for any categorical data without a specified contrast.
+
+abstract AbstractContrast
+
+termnames(term::Symbol, col::Any, contrast::AbstractContrast) =
+    ["$term - $name" for name in contrast.termnames]
+
+function cols(v::PooledDataVector, contrast::AbstractContrast)
+    ## make sure the levels of the contrast matrix and the categorical data
+    ## are the same by constructing a re-indexing vector. Indexing into
+    ## reindex with v.refs will give the corresponding row number of the
+    ## contrast matrix
+    reindex = [findfirst(l .== contrast.levels) for l in levels(v)]
+
+    ## TODO: add kwarg for full-rank contrasts (e.g., when intercept isn't
+    ## specified in a model frame).
+
+    return contrast.matrix[reindex[v.refs], :]
+end
+
+## Default contrast is treatment coding:
+cols(v::PooledDataVector) = cols(v, TreatmentContrast(v))
+termnames(term::Symbol, col::PooledDataArray) = termnames(term, col, TreatmentContrast(col))
+
+
+## Constructing a contrast from a non-pooled data vector will first pool it
+## 
+## (NOT SURE THIS IS GOOD: constructing columns also depends on levels, so for
+## _that_ to work would need to somehow hold onto pooled data...)
```

Comment:
Right. So it would make sense to specify the contrasts directly in the formula rather than in the call to `ModelFrame`. In R, the `C` function sets the contrasts as an attribute on factors (equivalent to PDAs), and as a standard function it can be used either separately, or directly in the formula. This solution implies storing an additional field for each PDA. Though it should be negligible if we only store a reference to the contrast type (and possibly the reference level), so it might be suited for the `CategoricalVariable` replacement for PDAs. An alternative would be to create a wrapper around a vector, which would hold both the data and the contrasts.

An advantage of these two strategies is that you would be able to specify the contrasts for a given variable once for all, which is a quite common need AFAICT. And the point about specifying the base level before the data is available would go away.